script: optimize `Element::create` for uncustomized built-in elements

mukilan opened the pull request:
The investigation in #37745 revealed that `Element::create` spends ~39%
of its runtime in `set_custom_element_state`. For uncustomized built-in
elements, this overhead is unnecessary as according to the spec, they
are equivalent to being initialized in the "uncustomized" state.
Although our rust implemention also initializes the elements in
"uncustomized" state, this particular call to `set_custom_element_state`
can't simply be removed as this call also ensures that the element's
DEFINED state is set to 'true', to match the `:defined` selector, based
on the custom element state.

So, introduce a new method that will only set the
`ElementState::DEFINED` flag by manipulating the state directly and call
it from `Element::create` when creating uncustomized, built-ins.

For more information about the peformance improvements, please refer to
[this comment][1].

[1]: https://github.com/servo/servo/issues/37745#issuecomment-3305477287

Testing: There should be no functional change, so this is covered by WPT suite.

mukilan committed:
script: optimize `Element::create` for uncustomized built-in elements

The investigation in #37745 revealed that `Element::create` spends ~39%
of its runtime in `set_custom_element_state`. For uncustomized built-in
elements, this overhead is unnecessary as according to the spec, they
are equivalent to being initialized in the "uncustomized" state.
Although our rust implemention also initializes the elements in
"uncustomized" state, this particular call to `set_custom_element_state`
can't simply be removed as this call also ensures that the element's
DEFINED state is set to 'true', to match the `:defined` selector, based
on the custom element state.

So, introduce a new method that will only set the
`ElementState::DEFINED` flag by manipulating the state directly and call
it from `Element::create` when creating uncustomized, built-ins.

For more information about the peformance improvements, please refer to
[this comment][1].

[1]: https://github.com/servo/servo/issues/37745#issuecomment-3305477287

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

servo-highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify script code, but no tests are modified. Please consider adding a test!

sagudev commented on components/script/dom/create.rs:
```diff
@@ -243,7 +243,9 @@ fn create_html_element(
-                result.set_custom_element_state(CustomElementState::Uncustomized);
+                // Note: This is a performance optimization. See the doc comment of the method for
+                // more information.
+                result.init_state_for_uncustomized_builtin_element();
```
Why not just changing set_custom_element_state? I think https://github.com/servo/servo/blob/af7de5ccf141c7cbb8a3792faeef916cdfa3dfc7/components/script/dom/element.rs#L433-L434 should use `&&`instead of `||` (per comment above), which should match the behavior of init_state_for_uncustomized_builtin_element.

mukilan replied:
Hmm...I think I might have overlooked something. The bottleneck is `ensure_pending_restyle` which will be called even from the `set_state` in `init_state_for_uncustomized_builtin_element` as the initial state would be empty. It should instead directly manipulate `self.state`. Not sure how the benchmarks show consistent difference, but I'll check this again after updating to new logic.

mrobinson replied:
The change to the microbenchmark probably makes this change worth it, even if it doesn't improve Speedometer.

mukilan replied:
I've added a new change [072b840](https://github.com/servo/servo/pull/39370/commits/072b84030a4d354e329b18d0f178421f9999f856) to remove the `self.rare_data.is_some()` check in `set_custom_element_state`. This is a redundant condition because, for uncustomized elements, the rare data will not exist. This is assuming an element will never go from non-uncustomized state to an uncustomized state. I can remove this if that is not a valid assumption to make.

mukilan committed:
script: remove redundant condition in set_custom_elememt_state

The rare data of an uncustomized element doesn't need to be inflated
assuming we never switch from non-uncustomized state to uncustomized.

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>
